types/src/v0/impls/l1.rs

```diff
@@ -64,180 +75,470 @@ impl L1BlockInfo {
-                sleep(retry).await;
-                continue;
-            };
-            break L1BlockInfo {
-                number,
-                hash,
-                timestamp: block.timestamp,
-            };
+                tracing::debug!(number, head, "waiting for L1 block");
+            }
+
+            // This should not happen: the event stream ended. All we can do is try again.
+            tracing::warn!(number, "L1 event stream ended unexpectedly; retry");
+            sleep(self.retry_delay).await;
         }
     }
+
     /// Get information about the given block.
     ///
     /// If the desired block number is not finalized yet, this function will block until it becomes
     /// finalized.
     pub async fn wait_for_finalized_block(&self, number: u64) -> L1BlockInfo {
-        // First just wait for the block to exist. This uses an efficient polling mechanism that
-        // polls more frequently as we get closer to expecting the block to be ready.
-        self.wait_for_block(number).await;
-
-        // Wait for the block to finalize. Since we know the block at least exists, we don't expect
-        // to have to wait _too_ long for it to finalize, so we poll relatively frequently, using
-        // the retry delay instead of the provider interval.
-        let finalized = loop {
-            let Some(block) = self.get_finalized_block().await else {
-                tracing::info!("waiting for finalized block");
-                sleep(self.retry_delay).await;
-                continue;
-            };
-            if block.number >= number {
-                break block;
+        loop {
+            // Subscribe to events before checking the current state, to ensure we don't miss a relevant
+            // event.
+            let mut events = self.receiver.activate_cloned();
+
+            // Check if the block we are waiting for already exists.
+            {
+                let state = self.state.lock().await;
+                if let Some(finalized) = state.snapshot.finalized {
+                    if finalized.number >= number {
+                        return self.get_finalized_block(state, number).await.1;
+                    }
+                    tracing::info!(
+                        number,
+                        finalized = ?state.snapshot.finalized,
+                        "waiting for l1 finalized block",
+                    );
+                };
             }
-            tracing::info!(current_finalized = %block.number, "waiting for finalized block");
-            sleep(self.retry_delay).await;
-            continue;
-        };
 
-        if finalized.number == number {
-            return finalized;
-        }
+            // Wait for the block.
+            while let Some(event) = events.next().await {
+                let L1Event::NewFinalized { finalized } = event else {
+                    continue;
+                };
+                if finalized.number < number {
```

Comment:
nit: for the latest block we do the comparison the other way around above: `if head >= number {` I think we can do it the same way both times. Then we also don't need this `continue`, I think.